src/mpi/datatype/typerep/src/typerep_yaksa_pack.c

```diff
@@ -499,235 +415,215 @@ int MPIR_Typerep_wait(MPIR_Typerep_req typerep_req)
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIR_TYPEREP_OP);
 
-    /* first Judge if source buffer is empty */
-    if (op == MPI_NO_OP)
-        is_empty_source = TRUE;
+    /* error checking */
+    MPIR_Assert(HANDLE_IS_BUILTIN(op));
+    MPIR_Assert(MPIR_DATATYPE_IS_PREDEFINED(source_dtp));
 
-    if (is_empty_source == FALSE) {
-        MPIR_Assert(MPIR_DATATYPE_IS_PREDEFINED(source_dtp));
-        MPIR_Datatype_get_size_macro(source_dtp, source_dtp_size);
-        MPIR_Datatype_get_extent_macro(source_dtp, source_dtp_extent);
+    mpi_errno = (*MPIR_OP_HDL_TO_DTYPE_FN(op)) (source_dtp);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    /* trivial cases */
+    if (op == MPI_NO_OP) {
+        goto fn_exit;
     }
 
-    if ((HANDLE_IS_BUILTIN(op))
-        && ((*MPIR_OP_HDL_TO_DTYPE_FN(op)) (source_dtp) == MPI_SUCCESS)) {
-        is_supported = MPIR_Typerep_reduce_is_supported(op, source_dtp);
+    bool use_yaksa = MPIR_Typerep_reduce_is_supported(op, source_dtp);
+
+    if (use_yaksa) {
+        mpi_errno = typerep_op_internal(source_buf, source_count, source_dtp,
+                                        target_buf, target_count, target_dtp, op, true);
     } else {
-        /* --BEGIN ERROR HANDLING-- */
-        mpi_errno = MPIR_Err_create_code(MPI_SUCCESS, MPIR_ERR_RECOVERABLE,
-                                         __func__, __LINE__, MPI_ERR_OP,
-                                         "**opnotpredefined", "**opnotpredefined %d", op);
-        return mpi_errno;
-        /* --END ERROR HANDLING-- */
+        /* swap host buffers if necessary */
+        void *save_targetbuf = NULL;
+        void *host = MPIR_gpu_host_swap(target_buf, target_count, target_dtp);
+        if (host) {
+            save_targetbuf = target_buf;
+            target_buf = host;
+        }
+
+        mpi_errno = typerep_op_internal(source_buf, source_count, source_dtp,
+                                        target_buf, target_count, target_dtp, op, false);
+
+        if (save_targetbuf) {
+            MPIR_gpu_swap_back(target_buf, save_targetbuf, target_count, target_dtp);
+            target_buf = save_targetbuf;
+        }
+    }
+    MPIR_ERR_CHECK(mpi_errno);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIR_TYPEREP_OP);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+static int typerep_op_basic(void *source_buf, void *target_buf, MPI_Aint count,
+                            MPI_Datatype datatype, MPI_Op op, bool use_yaksa)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    if (use_yaksa) {
+        yaksa_type_t yaksa_type = MPII_Typerep_get_yaksa_type(datatype);
+        yaksa_op_t yaksa_op = MPII_Typerep_get_yaksa_op(op);
+
+        MPI_Aint data_sz;
+        MPIR_Datatype_get_size_macro(datatype, data_sz);
+        data_sz *= count;
+
+        uintptr_t actual_pack_bytes;
+        yaksa_request_t request;
+        int rc;
+        rc = yaksa_ipack(source_buf, count, yaksa_type, 0, target_buf, data_sz, &actual_pack_bytes,
```

Comment:
I would prefer to have the `op_pack()` and `op_unpack()` functions.

Reply:
Let's add them back when we need them. With the IOV code, we definitely want to (as TODO) optimize it by issuing batch `ipack` then wait for them all. The `op_pack` will be a wrong abstraction that prevents it.

Does yaksa `iunpack` take care of non-contig datatype? If so, why do we need go through the IOVs?

Reply:
`iunpack` take a contiguous buffer as input and then produces a non-contig one.